rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,524 @@
+connector changes, and other forms of cluster sabotage.
+
+Adding an MFA restriction to administrative actions limits this attack vector
+by re-verifying the user's identity promptly before performing any administrative
+action.
+
+## Details
+
+### UX
+
+The UX of this feature will be very similar to Per-session MFA for `tsh`, `tctl`,
+and the WebUI, and Teleport Connect.
+
+When a user performs an admin action, they will be prompted to tap their MFA key.
+Each admin action will require a single unique tap or OTP code.
+
+```console
+$ tctl rm roles/access
+Tap any security key
+# success
+```
+
+#### Editing resources
+
+With `tctl edit` or the WebUI, it is possible to edit existing resources in a
+text editor. In this case, MFA will be prompted before proceeding to the text
+editor to prevent data loss from a failed MFA challenge.
+
+```console
+$ tctl edit roles/access
+Tap any security key
+# edit role in Vim
+# success
+```
+
+### Server configuration
+
+The requirement for MFA on admin actions will be made configurable through
+`cluster_auth_preference.admin_action_mfa: off | on`.
+
+When set to on, MFA will be required for admin actions for all users.
+
+When set to off, MFA will only be required for actions which already require MFA.
+
+This value will default to on/off based on the value of `cluster_auth_preference.second_factor`:
+
+- `second_factor = off | optional` => `admin_action_mfa = off`
+- `second_factor = on | otp | webauthn` => `admin_action_mfa = on`
+
+#### Backward Compatibility
+
+Once the Auth server begins to require MFA for admin actions, old clients with
+no way of providing MFA verification will fail to carry out admin actions.
+
+In order to maintain backwards compatibility between old clients and an upgraded
+server, we can't actually default `admin_action_mfa = on` as planned above.
+
+Instead, in the first major version of this feature (likely Teleport 14), it will
+always default to `off`. In the next major version (Teleport 15), the defaults
+above will be used.
+
+#### Built-in Roles
+
+Built-in roles will not require MFA to complete admin actions. This includes
+the `Admin` role which is used by `tctl` when performing operations directly
+on the Auth Server.
+
+#### Automated Use Cases
+
+In order to support automated use cases such as the
+[Teleport Terraform Provider](https://goteleport.com/docs/management/guides/terraform-provider/)
+or [Access Request Plugins](https://goteleport.com/docs/access-controls/access-request-plugins/),
+we need to provide a way for specific users to bypass MFA for admin actions.
+
+We will introduce a new role option, `robot`, to bypass MFA requirements.
+At first, this will only bypass the MFA requirement for admin actions, but
+could be reused to bypass other features which are not compatible with
+automated use cases.
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: access-request-plugin
+spec:
+  options:
+    robot: true
```

Comment:
Can we remove this extra knob as well? Certs issued by MachineID's non-interactive users should not require per session MFA for administrative actions if allowed by roles. All interactive clients should either opportunistically require in v14 or mandatory require in v15.

Reply:
> Certs issued by MachineID's non-interactive users should not require per session MFA for administrative actions if allowed by roles.

Great idea. We can skip MFA for impersonated certificates by checking the `impersonator` certificate extension for a built in role like `Bot` or `Admin`. 

Bots also require `create` permissions for `roles`, `users`, and `tokens`, making this method a bit more secure than a basic role option. @jentfoo 

We'll also need to update our plugin/API guides to use either MachineID, which seems to be our plan regardless:
```
> tctl auth sign --user=dev 
Generating credentials to allow a machine access to Teleport? We recommend Teleport's Machine ID! Find out more at https://goteleport.com/r/machineid-tip
...
```